parser-typechecker/src/Unison/Typechecker/Context.hs

```diff
@@ -521,7 +530,18 @@ markThenRetract hint body =
   markThenCallWithRetract hint \retract -> adjustNotes do
     r <- body
     ctx <- retract
-    pure ((r, ctx), substituteSolved ctx)
+    let solvedCtx = substituteSolved ctx
+    for_ ctx \case
+      var@(Ann v typ) -> do
+        Debug.debugM Debug.Temp "Ann" var
+        noteVarBinding v  (TypeVar.lowerType typ)
+      v@(Var{}) ->
+        Debug.debugM Debug.Temp "Var" v
+      (Solved _ v t) -> do
+        Debug.debugM Debug.Temp "Solved" v
+        noteVarBinding v (TypeVar.lowerType $ Type.getPolytype t)
+      _ -> pure ()
+    pure ((r, ctx), solvedCtx)
```

Comment:
This doesn't seem right to me... or it might not be wrong but could lead to quadratic performance.

I think you want to emit notes when things are solved. This seems like it will be traversing the same context over and over again.